tests/withBlockTracker.ts

```diff
@@ -0,0 +1,215 @@
+import util from 'util';
+import SafeEventEmitter from '@metamask/safe-event-emitter';
+import { JsonRpcRequest, JsonRpcResponse } from 'json-rpc-engine';
+import { Provider } from '../src/BaseBlockTracker';
+import {
+  SubscribeBlockTracker,
+  SubscribeBlockTrackerOptions,
+} from '../src/SubscribeBlockTracker';
+import {
+  PollingBlockTracker,
+  PollingBlockTrackerOptions,
+} from '../src/PollingBlockTracker';
+
+type BlockTracker = PollingBlockTracker | SubscribeBlockTracker;
+type BlockTrackerOptions<T> = T extends typeof PollingBlockTracker
+  ? PollingBlockTrackerOptions
+  : T extends typeof SubscribeBlockTracker
+  ? SubscribeBlockTrackerOptions
+  : never;
+type BlockTrackerConstructor<T extends BlockTracker> = new (
+  options?: BlockTrackerOptions<T>,
+) => T;
+
+type FakeProviderStub =
+  | {
+      methodName: string;
+      response: { result: any } | { error: string };
+    }
+  | {
+      methodName: string;
+      implementation: () => void;
+    }
+  | {
+      methodName: string;
+      error: string;
+    };
+
+export type WithBlockTrackerCallback<T extends BlockTracker> = (args: {
+  provider: FakeProvider;
+  blockTracker: T;
+}) => void | Promise<void>;
+
+interface FakeProviderOptions {
+  stubs?: FakeProviderStub[];
+}
+
+export interface WithBlockTrackerOptions<T extends BlockTracker> {
+  blockTracker?: {
+    [K in keyof BlockTrackerOptions<T>]?: BlockTrackerOptions<T>[K];
+  };
+  provider?: FakeProviderOptions;
+}
+
+/**
+ * FakeProvider is an implementation of the provider that a subclass of
+ * BaseBlockTracker takes, supporting the same expected methods with the same
+ * expected interface, except that fake responses for the various RPC methods
+ * that the provider supports can be supplied. This ends up easier to define
+ * than using `jest.spyOn(...).mockImplementationOnce(...)` due to the types
+ * that `sendAsync` takes.
+ */
+class FakeProvider extends SafeEventEmitter implements Provider {
+  private stubs: FakeProviderStub[];
+
+  private originalStubs: FakeProviderStub[];
+
+  constructor({ stubs = [] }: FakeProviderOptions = {}) {
+    super();
+    this.stubs = this.buildStubsFrom(stubs);
+    this.originalStubs = this.stubs.slice();
+  }
+
+  sendAsync<T, U>(
+    request: JsonRpcRequest<T>,
+    callback: (err: Error, response: JsonRpcResponse<U>) => void,
+  ) {
+    const index = this.stubs.findIndex(
+      (stub) => stub.methodName === request.method,
+    );
+
+    if (index !== -1) {
+      const stub = this.stubs[index];
+      this.stubs.splice(index, 1);
+      if ('implementation' in stub) {
+        stub.implementation();
+      } else if ('response' in stub) {
+        if ('result' in stub.response) {
+          callback(null as unknown as Error, {
```

Comment:
My thought is that it would end up being an overload in this case:

``` typescript
// Anything except null or undefined
type ErrorLike = string | number | symbol | boolean | object;

type SendAsyncCallback<T> = {
  (error: null, response: JsonRpcResponse<T>) => void;
  (error: ErrorLike, response: null) => void;
}

export interface Provider extends SafeEventEmitter {
  sendAsync: <T, U>(
    req: JsonRpcRequest<T>,
    cb: SendAsyncCallback<U>,
  ) => void;
}
```

I'm a bit worried about the downstream effects of this, though, so I was considering doing this in a separate PR, but if you don't think it would be a big deal, I can make it here.

Reply:
No that's OK, separate PR sounds perfectly fine